### PR TITLE
Global dependency graph

### PR DESCRIPTION
**Pull request by oheger-bosch**
Change the DependencyGraph format to be global for a package manager result, i.e. combine the dependencies of multiple projects.

Note: The new approach is not that different from the old one, but there is some shift in the responsibility of dependency information: The dependency graph is no longer maintained by a project, but projects only have references to the shared dependency graph. This also affects the conversion of an `OrtResult` from the graph format to the classic structure with project scopes, which is expected by the components that operate on dependency data.

The intension of this PR is to keep the single changes small without breaking any tests. To achieve this, some redundancy had to be introduced temporarily. Especially the `Project` class is added a new dependency-related property, and the obsolete one is removed only later after other components have been reworked. If this is not desired, some commits can be squashed, but then there is a rather big bang change.

### CODE REVIEW

File: analyzer/src/funTest/assets/projects/external/multi-kotlin-project-expected-output-cli.yml [hunk 1]
```diff
@@ -15,95 +15,222 @@ project:
...
-    - "Maven:org.jetbrains.kotlin:kotlin-stdlib-common:1.3.11"
-    - "Maven:org.jetbrains:annotations:13.0"
-    - "Maven:org.jetbrains.kotlin:kotlin-compiler-embeddable:1.3.11"
-    - "Maven:org.jetbrains.kotlin:kotlin-script-runtime:1.3.11"
-    - "Maven:org.jetbrains.kotlin:kotlin-reflect:1.3.11"
-    - "Maven:org.jetbrains.kotlin:kotlin-scripting-compiler-embeddable:1.3.11"
-    scope_roots:
-    - dependencies:
-      - pkg: 1
+  scopes:
```
Comment (sschuberth) [inline analyzer/src/funTest/assets/projects/external/multi-kotlin-project-expected-output-cli.yml:18]: Hint:

> When support for the DependencyGraph format was added to Gradle in
c8d69ca, the expectations of tests
have been adapted to the new format. This is going to be problematic

The formatting looks odd here as GitHub shortens commit SHA1 to 7 chars. You can align the plain text and rendered view by also only using 7 chars in the original commit message.

File: model/src/main/kotlin/DependencyGraph.kt [hunk 2]
```diff
@@ -61,6 +61,30 @@ data class DependencyGraph(
      */
     val scopes: Map<String, List<RootDependencyIndex>>
 ) {
+    companion object {
```
Comment (sschuberth) [inline model/src/main/kotlin/DependencyGraph.kt:65]: Commit message nit:

> so that it can be determined, which scope belongs to which
project.

I believe there should be no comma in here.

**Review by sschuberth — CHANGES_REQUESTED** [review]
I only skimmed over the changes in this first round. Will need to make a more thorough follow-up review later.

File: analyzer/src/funTest/kotlin/managers/Extensions.kt [hunk 3]
```diff
@@ -31,12 +31,23 @@ import org.ossreviewtoolkit.model.yamlMapper
...
-        else result
+        if (resolveScopes) {
+            val resolvedProject = result.project.withResolvedScopes(managerResult.dependencyGraph)
+
+            // The result must contain packages for all the dependencies declared by the project; otherwise, the
+            // check in ProjectAnalyzerResult.init fails. When using a shared dependency graph, the set of packages
+            // is typically empty, so it has to be populated manually.
+            val packages = result.packages.takeIf { it.isNotEmpty() } ?: managerResult.sharedPackages
+            result.copy(project = resolvedProject, packages = packages.toSortedSet()
+            )
```
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply analyzer/src/funTest/kotlin/managers/Extensions.kt:50]: Doh! While at it, I extracted this block into a separate function. This will be useful for other tests for package managers as well.

File: model/src/main/kotlin/DependencyGraph.kt [hunk 4]
```diff
@@ -61,6 +61,30 @@ data class DependencyGraph(
...
+         * To achieve this, they are qualified with the project ID.
+         */
+        fun qualifyScope(projectId: Identifier, scopeName: String): String =
+            "${projectId.namespace}:${projectId.name}:${projectId.version}:$scopeName"
+
+        /**
+         * Extract the plain (un-qualified) scope name from the given qualified [scopeName]. If the passed in
+         * [scopeName] is not qualified, return it unchanged.
+         */
+        fun unQualifyScope(scopeName: String): String = scopeName.substringAfterLast(':')
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply model/src/main/kotlin/DependencyGraph.kt:86]: It should be as we have

https://github.com/oss-review-toolkit/ort/blob/274d7b2273ba9601c52ece0d7b3298e93c233ad1/model/src/main/kotlin/Identifier.kt#L83-L86

or?

[the remark this answers is not in the document]
Comment (oheger-bosch) [reply model/src/main/kotlin/DependencyGraph.kt:86]: Yes, I was using `Identifier` as a model and was hoping the same.

[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply model/src/main/kotlin/DependencyGraph.kt:86]: The code above doesn't help because the scope is not part of the identifier. But I think the risk pretty low that scope names contain ":". But anyway, you could change that to return the string after the third ":" to be on the safe side.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/OrtResult.kt [hunk 5]
```diff
@@ -93,6 +94,13 @@ data class OrtResult(
             evaluator = null,
             labels = emptyMap()
         )
+
+        /**
+         * Load an [OrtResult] from the given [file]. This function deserializes the result from the specified input
+         * file and ensures that it has been properly initialized. It therefore should be always used rather than
+         * doing the deserialization manually.
+         */
+        fun load(file: File): OrtResult = file.readValue<OrtResult>().withResolvedScopes()
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/OrtResult.kt:103]: I think this could also have been solved by providing a custom deserializer for `OrtResult` which calls the function after deserialization.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/AnalyzerResultBuilder.kt [hunk 6]
```diff
@@ -61,23 +64,41 @@ class AnalyzerResultBuilder(private val curationProvider: PackageCurationProvide
             issues[existingProject.id] = projectIssues + issue
         } else {
             projects += projectAnalyzerResult.project
+            addPackages(projectAnalyzerResult.packages)
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/AnalyzerResultBuilder.kt:67]: The refactoring of `addResult` seems unrelated to the change and should have been done in a separate commit.

File: analyzer/src/funTest/kotlin/managers/Extensions.kt [hunk 7]
```diff
@@ -26,17 +26,35 @@ import io.kotest.matchers.should
...
+import org.ossreviewtoolkit.analyzer.PackageManagerResult
+import org.ossreviewtoolkit.model.DependencyGraph
 import org.ossreviewtoolkit.model.ProjectAnalyzerResult
 import org.ossreviewtoolkit.model.yamlMapper
 
 fun Any?.toYaml() = yamlMapper.writeValueAsString(this)!!
 
-fun PackageManager.resolveSingleProject(definitionFile: File, resolveScopes: Boolean = false): ProjectAnalyzerResult =
-    resolveDependencies(listOf(definitionFile)).projectResults[definitionFile].let { resultList ->
+fun PackageManager.resolveSingleProject(definitionFile: File, resolveScopes: Boolean = false): ProjectAnalyzerResult {
```
Comment (mnonnenmacher) [inline analyzer/src/funTest/kotlin/managers/Extensions.kt:36]: The refactoring of `resolveSingleProject` seems unrelated to the change and should have been done in a separate commit.

File: analyzer/src/main/kotlin/Analyzer.kt [hunk 8]
```diff
@@ -135,17 +135,22 @@ class Analyzer(private val config: AnalyzerConfiguration) {
...
                             "Projects $projectString must be of type '${manager.managerName}'."
                         }
                     }
+
+                    manager to results
                 }
             }.forEach { resolutionResult ->
-                resolutionResult.await().forEach { (_, analyzerResults) ->
-                    analyzerResults.forEach { analyzerResultBuilder.addResult(it) }
+                val (manager, results) = resolutionResult.await()
```
Comment (mnonnenmacher) [inline analyzer/src/main/kotlin/Analyzer.kt:149]: `resolutionResult` should be renamed because `ResolutionResult` is the type of the variable now called `results` which is confusing.

File: model/src/test/kotlin/ProjectTest.kt [hunk 9]
```diff
@@ -101,18 +86,23 @@ private fun createDependencyGraph(qualified: Boolean = false): DependencyGraph {
...
+            shouldThrow<IllegalArgumentException> {
+                Project(
+                    id = projectId,
+                    definitionFilePath = "/some/path/pom.xml",
+                    declaredLicenses = sortedSetOf(),
+                    vcs = VcsInfo.EMPTY,
+                    homepageUrl = "https://test-project.example.org/home.html",
+                    scopeDependencies = sortedSetOf(mockk()),
+                    scopeNames = sortedSetOf("test", "compile", "other")
+                )
```
Comment (mnonnenmacher) [inline model/src/test/kotlin/ProjectTest.kt:101]: Could use `Project.EMPTY.copy()` to provide only the required values.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/DependencyGraph.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply model/src/main/kotlin/DependencyGraph.kt:86]: The function has been renamed and now returns the part after the 3rd separator.

File: model/src/main/kotlin/DependencyGraph.kt [hunk 10]
```diff
@@ -85,21 +86,42 @@ data class DependencyGraph(
         fun unQualifyScope(scopeName: String): String = scopeName.substringAfterLast(':')
     }
 
+    /**
+     * Stores a mapping from dependency indices to [PackageReference] objects. This is needed when converting the
+     * data of this object to the classical layout of dependency information. The structure is created once and can
+     * then be used to convert parts of this graph.
```
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply model/src/main/kotlin/DependencyGraph.kt:92]: Slightly reworded.

File: analyzer/src/main/kotlin/AnalyzerResultBuilder.kt (hunk 6 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/AnalyzerResultBuilder.kt:67]: Split to a separate commit.

File: analyzer/src/funTest/kotlin/managers/Extensions.kt (hunk 7 above)
Comment (oheger-bosch) [reply analyzer/src/funTest/kotlin/managers/Extensions.kt:36]: I think, this has to be done in the same commit together with the changes on the test classes; otherwise, tests will fail.

File: analyzer/src/main/kotlin/Analyzer.kt (hunk 8 above)
Comment (oheger-bosch) [reply analyzer/src/main/kotlin/Analyzer.kt:149]: I renamed `resolutionResult` to `deferredResult` and `results` to `managerResult` (as it is of type `PackageManagerResult`).

**Review by mnonnenmacher — DISMISSED** [review]

File: model/src/main/kotlin/DependencyGraph.kt [hunk 11]
```diff
@@ -61,6 +61,39 @@ data class DependencyGraph(
...
+            "${projectId.namespace}:${projectId.name}:${projectId.version}:$scopeName"
+
+        /**
+         * Extract the plain (un-qualified) scope name from the given qualified [scopeName]. If the passed in
+         * [scopeName] is not qualified, return it unchanged.
+         */
+        fun unqualifyScope(scopeName: String): String =
+            // To handle the case that the scope contains the separator character, cut off the parts for the
+            // namespace, the name, and the version.
+            scopeName.substringAfterIth(':', 3)
```
Comment (sschuberth) [inline model/src/main/kotlin/DependencyGraph.kt:88]: Can't we simply use `scopeName.substringAfterLast(':')` here (which has a default of `missingDelimiterValue: String = this`)? Or `scopeName.split(':').getOrElse(4) { scopeName }`? I'd really like to get rid of the overly specific `substringAfterIth` function 😁

File: model/src/main/kotlin/Project.kt [hunk 12]
```diff
@@ -99,9 +99,18 @@ data class Project(
      * Contains dependency information as a [DependencyGraph]. This is an alternative format to store the dependencies
      * referenced by the various scopes. Use the [scopes] property to access dependency information independent on
      * the concrete representation.
+     * TODO: Remove this after all affected package managers have been converted to use a shared graph.
```
Comment (sschuberth) [inline model/src/main/kotlin/Project.kt]: Nit: Maybe leave an empty line to separate the TODO from the docs, to make it stand out more?

File: model/src/main/kotlin/DependencyGraph.kt (hunk 11 above)
Comment (oheger-bosch) [reply model/src/main/kotlin/DependencyGraph.kt:88]: Well, as @mnonnenmacher has correctly pointed out, there is a (small) risk that a scope name contains the separator character. So a robust solution really needs to obtain the substring after the 3rd occurrence of this separator. If I am not missing something, none of the simple solutions does this.

Comment (sschuberth) [reply model/src/main/kotlin/DependencyGraph.kt:88]: Wouldn't passing a limit of 4 to split() cope with that?

Comment (oheger-bosch) [reply model/src/main/kotlin/DependencyGraph.kt:88]: Yes, this does the trick. Nice.

File: cli/src/funTest/kotlin/OrtMainFunTest.kt [hunk 13]
```diff
@@ -160,6 +166,23 @@ class OrtMainFunTest : StringSpec() {
...
+
+            val result = cmd.readOrtResult(resultFile)
+
+            result.analyzer?.result shouldNotBeNull {
+                dependencyGraphs.keys should beEmpty()
+            }
+
+            val project = result.getProject(Identifier("Maven:com.vdurmont:semver4j:3.1.0"))
+
+            project.shouldNotBeNull {
```
Comment (sschuberth) [inline cli/src/funTest/kotlin/OrtMainFunTest.kt:182]: Nit: Ideally, this should also use the infix notation.

**Review by sschuberth — DISMISSED** [review]

File: cli/src/funTest/kotlin/OrtMainFunTest.kt (hunk 13 above)
Comment (sschuberth) [reply cli/src/funTest/kotlin/OrtMainFunTest.kt:182]: @oheger-bosch looks like you need to rebase anyway to get rid of CI failures, and if doing so, would you mind to address my nit despite the approval?

**Review by sschuberth — APPROVED** [review]
